metaflow/plugins/aws/awslambda/lambda_runner.py

```diff
@@ -0,0 +1,314 @@
+    DATATOOLS_S3ROOT,
+    DATASTORE_SYSROOT_S3,
+)
+
+
+class LogTailer(Thread):
+
+    _stop_deadline: Optional[float] = None
+    _log_end_token: Optional[str] = None
+
+    def __init__(self, log_group, log_stream, echo, log_start_token):
+        """
+        Parameters
+        ----------
+        log_group : str
+            Cloudwatch log group name
+        log_stream : str
+            Cloudwatch log stream name
+        echo : Callable[str]
+            Printer function to call for each log line
+        log_start_token : str
+            The tailer will skip logs until log_start_token is first seen. The reason
+            is that some of the "preamble" shell commands can be verbose, e.g. tar
+            likes to complain about timestamps too far in the past. These logs will not
+            be printed to the user (but they will still be available in Cloudwatch for
+            debugging).
+        """
+        self._log_group = log_group
+        self._log_stream = log_stream
+        self._echo = echo
+        self._log_start_token = log_start_token
+        super(LogTailer, self).__init__()
+        self.daemon = True
+
+    def stop_tail(self, timeout_seconds, log_end_token):
+        """
+        Stop tailing.
+
+        After Lambda invocation is done, there is generally no way for us to know if
+        there are any log messages that were written by Lambda to Cloudwatch, but
+        haven't yet been read by this LogTailer.
+
+        So we use the following strategy to decide when to stop reading: Lambda will
+        write a unique log_end_token at the end. On the reading end, after lambda
+        finishes, we'll read until we encounter that token, or until timeout_second
+        expires. The timeout is useful for the case where Lambda dies unexpectedly and
+        doesn't get a chance to write the token.
+        """
+        # Slightly awkward naming here to make sure there is no conflict with Thread
+        # internal members.
+        self._stop_deadline = time.time() + timeout_seconds
+        self._log_end_token = log_end_token
+
+    def run(self):
+        logs = BatchLogs(self._log_group, self._log_stream, sleep_on_no_data=1)
+        seen_start_token = False
+        while True:
+            for line in logs:
+                if line:
+                    if self._log_end_token is not None and line == self._log_end_token:
+                        return
+                    else:
+                        if (
+                            not seen_start_token
+                            and line.strip() == self._log_start_token
+                        ):
+                            seen_start_token = True
+                        elif seen_start_token:
+                            self._echo(line)
+                else:
+                    if self._stop_deadline and self._stop_deadline < time.time():
+                        return
+
+
+def lambda_name(flow_name, step_name):
```

Comment:
What if we constructed the `lambda_name` to be `flow_name-run_id-step_name-task_id-attempt_id` instead? That will also negate the need for introducing `force_fetch_image`, I believe?

Reply:
Yes it would, but it would add 5-30 seconds of extra latency to create the Lambda for each task. One of the goals of this PR was to add a way to run your code in the cloud with minimal latency so it feels delightful, so I was optimizing for lower startup time whenever I could here.

Reply:
We would have to necessarily isolate the `lambda` to at least the `run_id` which would mean that there is no specific need for `force_fetch_image`.